File: app/src/main/java/org/mozilla/focus/navigation/TransactionHelper.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.navigation;

import android.arch.lifecycle.DefaultLifecycleObserver;
import android.arch.lifecycle.LifecycleOwner;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.text.TextUtils;
import android.view.animation.Animation;

import org.mozilla.focus.R;

import static org.mozilla.focus.navigation.ScreenNavigator.BrowserScreen;
import static org.mozilla.focus.navigation.ScreenNavigator.FIRST_RUN_FRAGMENT_TAG;
import static org.mozilla.focus.navigation.ScreenNavigator.HOME_FRAGMENT_TAG;
import static org.mozilla.focus.navigation.ScreenNavigator.HomeScreen;
import static org.mozilla.focus.navigation.ScreenNavigator.URL_INPUT_FRAGMENT_TAG;

class TransactionHelper implements DefaultLifecycleObserver {
    private final ScreenNavigator.HostActivity activity;
    private BackStackListener backStackListener;

    private static final String ENTRY_TAG_SEPARATOR = "#";

    TransactionHelper(@NonNull ScreenNavigator.HostActivity activity) {
        this.activity = activity;
        this.backStackListener = new BackStackListener(this);
    }

    @Override
    public void onStart(@NonNull LifecycleOwner owner) {
        this.activity.getSupportFragmentManager()
                .addOnBackStackChangedListener(this.backStackListener = new BackStackListener(this));
    }

    @Override
    public void onStop(@NonNull LifecycleOwner owner) {
        this.activity.getSupportFragmentManager()
                .removeOnBackStackChangedListener(this.backStackListener);
        this.backStackListener.onStop();
    }

    void showHomeScreen(boolean animated, @EntryData.EntryType int type) {
        if (isStateSaved()) {
            return;
        }
        this.prepareHomeScreen(animated, type).commit();
        this.activity.getSupportFragmentManager().executePendingTransactions();
    }

    void showFirstRun() {
        if (isStateSaved()) {
            return;
        }
        this.prepareFirstRun().commit();
    }

    void showUrlInput(@Nullable String url, String sourceFragment) {
        if (isStateSaved()) {
            return;
        }
        final FragmentManager fragmentManager = this.activity.getSupportFragmentManager();
        final Fragment existingFragment = fragmentManager.findFragmentByTag(URL_INPUT_FRAGMENT_TAG);
        if (existingFragment != null && existingFragment.isAdded() && !existingFragment.isRemoving()) {
            // We are already showing an URL input fragment. This might have been a double click on the
            // fake URL bar. Just ignore it.
            return;
        }

        this.prepareUrlInput(url, sourceFragment)
                .addToBackStack(makeEntryTag(URL_INPUT_FRAGMENT_TAG, EntryData.TYPE_FLOATING))
                .commit();
    }

    void dismissUrlInput() {
        final FragmentManager mgr = this.activity.getSupportFragmentManager();
        if (mgr.isStateSaved()) {
            return;
        }
        mgr.popBackStack();
    }

    boolean shouldFinish() {
        FragmentManager manager = this.activity.getSupportFragmentManager();
        int entryCount = manager.getBackStackEntryCount();
        if (entryCount == 0) {
            return true;
        }

        FragmentManager.BackStackEntry lastEntry = manager.getBackStackEntryAt(entryCount - 1);
        return EntryData.TYPE_ROOT == getEntryType(lastEntry);
    }

    void popAllScreens() {
        FragmentManager manager = this.activity.getSupportFragmentManager();
        int entryCount = manager.getBackStackEntryCount();
        while (entryCount > 0) {
            if (!manager.isStateSaved()) {
                // Can not perform this action after onSaveInstanceState
                manager.popBackStack();
                entryCount--;
            }
        }
        manager.executePendingTransactions();
    }

    boolean popScreensUntil(@Nullable String targetEntryName, @EntryData.EntryType int type) {
        boolean clearAll = (targetEntryName == null);
        FragmentManager manager = this.activity.getSupportFragmentManager();
        int entryCount = manager.getBackStackEntryCount();
        boolean found = false;
        while (entryCount > 0) {
            FragmentManager.BackStackEntry entry = manager.getBackStackEntryAt(entryCount - 1);
            if (!clearAll
                    && TextUtils.equals(targetEntryName, getEntryTag(entry))
                    && type == getEntryType(entry)) {
                found = true;
                break;
            }
            manager.popBackStack();
            entryCount--;
        }
        manager.executePendingTransactions();
        return found;
    }

    @Nullable
    Fragment getLatestCommitFragment() {
        FragmentManager manager = this.activity.getSupportFragmentManager();
        int count = manager.getBackStackEntryCount();
        if (count == 0) {
            return null;
        }

        String tag = getFragmentTag(count - 1);
        return manager.findFragmentByTag(tag);
    }

    private FragmentTransaction prepareFirstRun() {
        final FragmentManager fragmentManager = this.activity.getSupportFragmentManager();
        final ScreenNavigator.Screen screen = this.activity.createFirstRunScreen();
        FragmentTransaction transaction = fragmentManager.beginTransaction();
        if (fragmentManager.findFragmentByTag(FIRST_RUN_FRAGMENT_TAG) == null) {
            transaction.replace(R.id.container, screen.getFragment(), FIRST_RUN_FRAGMENT_TAG)
                    .addToBackStack(makeEntryTag(FIRST_RUN_FRAGMENT_TAG, EntryData.TYPE_ROOT));
        }

        return transaction;
    }

    private FragmentTransaction prepareHomeScreen(boolean animated, @EntryData.EntryType int type) {
        final FragmentManager fragmentManager = this.activity.getSupportFragmentManager();
        final HomeScreen homeScreen = this.activity.createHomeScreen();

        FragmentTransaction transaction = fragmentManager.beginTransaction();
        int enterAnim = animated ? R.anim.tab_transition_fade_in : 0;
        int exitAnim = (type == EntryData.TYPE_ROOT) ? 0 : R.anim.tab_transition_fade_out;
        transaction.setCustomAnimations(enterAnim, 0, 0, exitAnim);

        transaction.add(R.id.container, homeScreen.getFragment(), HOME_FRAGMENT_TAG);
        transaction.addToBackStack(makeEntryTag(HOME_FRAGMENT_TAG, type));

        return transaction;
    }

    private FragmentTransaction prepareUrlInput(@Nullable String url, String parentFragmentTag) {
        final FragmentManager fragmentManager = this.activity.getSupportFragmentManager();
        final ScreenNavigator.Screen urlScreen = this.activity.createUrlInputScreen(url, parentFragmentTag);
        FragmentTransaction transaction = fragmentManager.beginTransaction();
        transaction.add(R.id.container, urlScreen.getFragment(), URL_INPUT_FRAGMENT_TAG);
        return transaction;
    }

    void onUrlInputScreenVisible(boolean visible) {
        final FragmentManager fragmentManager = this.activity.getSupportFragmentManager();
        final ScreenNavigator.Screen homeFragment =
                (ScreenNavigator.Screen) fragmentManager.findFragmentByTag(HOME_FRAGMENT_TAG);
        if (homeFragment != null && homeFragment.getFragment().isVisible()) {
            if (homeFragment instanceof HomeScreen) {
                ((HomeScreen) homeFragment).onUrlInputScreenVisible(visible);
            }
        }
    }

    private String getFragmentTag(int backStackIndex) {
        FragmentManager manager = this.activity.getSupportFragmentManager();
        return getEntryTag(manager.getBackStackEntryAt(backStackIndex));
    }

    private String getEntryTag(FragmentManager.BackStackEntry entry) {
        String result[] = entry.getName().split(ENTRY_TAG_SEPARATOR);
        return result[0];
    }

    private @EntryData.EntryType
    int getEntryType(FragmentManager.BackStackEntry entry) {
        String result[] = entry.getName().split(ENTRY_TAG_SEPARATOR);
        return Integer.parseInt(result[1]);
    }

    private String makeEntryTag(String tag, @EntryData.EntryType int type) {
        return tag + ENTRY_TAG_SEPARATOR + type;
    }

    private boolean isStateSaved() {
        FragmentManager manager = this.activity.getSupportFragmentManager();
        return manager == null || manager.isStateSaved();
    }

    private static class BackStackListener implements FragmentManager.OnBackStackChangedListener {
        private Runnable stateRunnable;
        private TransactionHelper helper;

        BackStackListener(TransactionHelper helper) {
            this.helper = helper;
        }

        @Override
        public void onBackStackChanged() {
            if (this.helper == null) {
                return;
            }

            FragmentManager manager = this.helper.activity.getSupportFragmentManager();
            Fragment fragment = manager.findFragmentById(R.id.browser);
            if (fragment instanceof BrowserScreen) {
                setBrowserState(shouldKeepBrowserRunning(this.helper), this.helper);
            }
        }

        private void onStop() {
            this.helper = null;
            this.stateRunnable = null;
        }

        private boolean shouldKeepBrowserRunning(@NonNull TransactionHelper helper) {
            FragmentManager manager = helper.activity.getSupportFragmentManager();
            int entryCount = manager.getBackStackEntryCount();
            for (int i = entryCount - 1; i >= 0; --i) {
                FragmentManager.BackStackEntry entry = manager.getBackStackEntryAt(i);
                if (helper.getEntryType(entry) != EntryData.TYPE_FLOATING) {
                    return false;
                }
            }
            return true;
        }

        private void setBrowserState(boolean isForeground, @NonNull TransactionHelper helper) {
            this.stateRunnable = () -> setBrowserForegroundState(isForeground);

            FragmentAnimationAccessor actor = getTopAnimationAccessibleFragment(helper);
            Animation anim;
            if (actor == null
                    || ((anim = actor.getCustomEnterTransition()) == null)
                    || anim.hasEnded()) {
                executeStateRunnable();
            } else {
                anim.setAnimationListener(new Animation.AnimationListener() {
                    @Override
                    public void onAnimationStart(Animation animation) {
                    }

                    @Override
                    public void onAnimationEnd(Animation animation) {
                        executeStateRunnable();
                    }

                    @Override
                    public void onAnimationRepeat(Animation animation) {
                    }
                });
            }
        }

        private void setBrowserForegroundState(boolean isForeground) {
            if (this.helper == null) {
                return;
            }

            FragmentManager manager = this.helper.activity.getSupportFragmentManager();
            BrowserScreen browserFragment = (BrowserScreen) manager.findFragmentById(R.id.browser);
            if (isForeground) {
                browserFragment.goForeground();
            } else {
                browserFragment.goBackground();
            }
        }

        private void executeStateRunnable() {
            if (this.stateRunnable != null) {
                this.stateRunnable.run();
                this.stateRunnable = null;
            }
        }

        @Nullable
        FragmentAnimationAccessor getTopAnimationAccessibleFragment(@NonNull TransactionHelper helper) {
            Fragment top = helper.getLatestCommitFragment();
            if (top != null && top instanceof FragmentAnimationAccessor) {
                return (FragmentAnimationAccessor) top;
            }
            return null;
        }
    }

    static class EntryData {
        /**
         * argument passed to {@link FragmentTransaction#addToBackStack(String)}, pressing back when this
         * type of fragment is in foreground will close the app
         */
        static final int TYPE_ROOT = 0;

        /**
         * argument passed to {@link FragmentTransaction#addToBackStack(String)}, adding fragment of
         * this type will make browser fragment go to background
         */
        static final int TYPE_ATTACHED = 1;

        /**
         * argument passed to {@link FragmentTransaction#addToBackStack(String)}, browsing fragment
         * will still be in foreground after adding this type of fragment.
         */
        static final int TYPE_FLOATING = 2;

        @IntDef({TYPE_ROOT, TYPE_ATTACHED, TYPE_FLOATING})
        @interface EntryType {
        }
    }
}

File: app/src/main/java/org/mozilla/focus/navigation/TransactionHelper.kt
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
package org.mozilla.focus.navigation

import android.text.TextUtils
import android.view.animation.Animation
import androidx.annotation.IntDef
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import org.mozilla.focus.R
import org.mozilla.focus.navigation.ScreenNavigator.BrowserScreen
import org.mozilla.focus.navigation.ScreenNavigator.HomeScreen
import org.mozilla.focus.navigation.ScreenNavigator.HostActivity
import org.mozilla.focus.navigation.TransactionHelper.EntryData.EntryType

internal class TransactionHelper(private val activity: HostActivity) : DefaultLifecycleObserver {
    private var backStackListener: BackStackListener? = null
    private val topFragmentState = MutableLiveData<TopFragmentState>()

    init {
        registerBackStackListener()
    }

    override fun onStart(owner: LifecycleOwner) {
        registerBackStackListener()
    }

    override fun onStop(owner: LifecycleOwner) {
        unregisterBackStackListener()
    }

    fun showHomeScreen(animated: Boolean, @EntryType type: Int, executeImmediatly: Boolean) {
        if (isStateSaved) {
            return
        }
        prepareHomeScreen(animated, type).commit()
        if (executeImmediatly) {
            activity.getSupportFragmentManager().executePendingTransactions()
        }
    }

    fun showFirstRun() {
        if (isStateSaved) {
            return
        }
        prepareFirstRun().commit()
    }

    fun showUrlInput(url: String?, sourceFragment: String) {
        if (isStateSaved) {
            return
        }
        val fragmentManager = activity.getSupportFragmentManager()
        val existingFragment = fragmentManager.findFragmentByTag(ScreenNavigator.URL_INPUT_FRAGMENT_TAG)
        if (existingFragment != null && existingFragment.isAdded && !existingFragment.isRemoving) {
            // We are already showing an URL input fragment. This might have been a double click on the
            // fake URL bar. Just ignore it.
            return
        }
        prepareUrlInput(url, sourceFragment)
            .addToBackStack(makeEntryTag(ScreenNavigator.URL_INPUT_FRAGMENT_TAG, EntryData.TYPE_FLOATING))
            .commit()
    }

    fun dismissUrlInput() {
        val mgr = activity.getSupportFragmentManager()
        if (mgr.isStateSaved) {
            return
        }
        mgr.popBackStack()
    }

    fun shouldFinish(): Boolean {
        val manager = activity.getSupportFragmentManager()
        val entryCount = manager.backStackEntryCount
        if (entryCount == 0) {
            return true
        }
        val lastEntry = manager.getBackStackEntryAt(entryCount - 1)
        return EntryData.TYPE_ROOT == getEntryType(lastEntry)
    }

    fun popAllScreens() {
        val manager = activity.getSupportFragmentManager()
        var entryCount = manager.backStackEntryCount
        while (entryCount > 0) {
            manager.popBackStack()
            entryCount--
        }
        manager.executePendingTransactions()
    }

    fun popScreensUntil(targetEntryName: String?, @EntryType type: Int, executeImmediately: Boolean): Boolean {
        val clearAll = targetEntryName == null
        val manager = activity.getSupportFragmentManager()
        var entryCount = manager.backStackEntryCount
        var found = false
        while (entryCount > 0) {
            val entry = manager.getBackStackEntryAt(entryCount - 1)
            if (!clearAll && TextUtils.equals(targetEntryName, getEntryTag(entry)) && type == getEntryType(entry)) {
                found = true
                break
            }
            manager.popBackStack()
            entryCount--
        }
        if (executeImmediately) {
            manager.executePendingTransactions()
        }
        return found
    }

    val latestCommitFragment: Fragment?
        get() {
            val manager = activity.getSupportFragmentManager()
            val count = manager.backStackEntryCount
            if (count == 0) {
                return null
            }
            val tag = getFragmentTag(count - 1)
            return manager.findFragmentByTag(tag)
        }

    fun getTopFragmentState(): LiveData<TopFragmentState> {
        return topFragmentState
    }

    fun executePendingTransaction() {
        activity.getSupportFragmentManager().executePendingTransactions()
    }

    private fun prepareFirstRun(): FragmentTransaction {
        val fragmentManager = activity.getSupportFragmentManager()
        val screen = activity.createFirstRunScreen()
        val transaction = fragmentManager.beginTransaction()
        if (fragmentManager.findFragmentByTag(ScreenNavigator.FIRST_RUN_FRAGMENT_TAG) == null) {
            transaction.replace(R.id.container, screen.getFragment(), ScreenNavigator.FIRST_RUN_FRAGMENT_TAG)
                .addToBackStack(makeEntryTag(ScreenNavigator.FIRST_RUN_FRAGMENT_TAG, EntryData.TYPE_ROOT))
        }
        return transaction
    }

    private fun prepareHomeScreen(animated: Boolean, @EntryType type: Int): FragmentTransaction {
        val fragmentManager = activity.getSupportFragmentManager()
        val homeScreen = activity.createHomeScreen()
        val transaction = fragmentManager.beginTransaction()
        val enterAnim = if (animated) R.anim.tab_transition_fade_in else 0
        val exitAnim = if (type == EntryData.TYPE_ROOT) 0 else R.anim.tab_transition_fade_out
        transaction.setCustomAnimations(enterAnim, 0, 0, exitAnim)
        transaction.add(R.id.container, homeScreen.getFragment(), ScreenNavigator.HOME_FRAGMENT_TAG)
        transaction.addToBackStack(makeEntryTag(ScreenNavigator.HOME_FRAGMENT_TAG, type))
        return transaction
    }

    private fun prepareUrlInput(url: String?, parentFragmentTag: String): FragmentTransaction {
        val fragmentManager = activity.getSupportFragmentManager()
        val urlScreen: ScreenNavigator.Screen = activity.createUrlInputScreen(url, parentFragmentTag)
        val transaction = fragmentManager.beginTransaction()
        transaction.add(R.id.container, urlScreen.getFragment(), ScreenNavigator.URL_INPUT_FRAGMENT_TAG)
        return transaction
    }

    fun onUrlInputScreenVisible(visible: Boolean) {
        val fragmentManager = activity.getSupportFragmentManager()
        val homeFragment = fragmentManager.findFragmentByTag(ScreenNavigator.HOME_FRAGMENT_TAG) as ScreenNavigator.Screen?
        if (homeFragment != null && homeFragment.getFragment().isVisible) {
            if (homeFragment is HomeScreen) {
                homeFragment.onUrlInputScreenVisible(visible)
            }
        }
    }

    private fun getFragmentTag(backStackIndex: Int): String {
        val manager = activity.getSupportFragmentManager()
        return getEntryTag(manager.getBackStackEntryAt(backStackIndex))
    }

    private fun getEntryTag(entry: FragmentManager.BackStackEntry): String {
        return entry.name?.let {
            it.split(ENTRY_TAG_SEPARATOR).toTypedArray()[0]
        } ?: ""
    }

    @EntryType
    private fun getEntryType(entry: FragmentManager.BackStackEntry): Int {
        return entry.name?.let {
            it.split(ENTRY_TAG_SEPARATOR).toTypedArray()[1].toInt()
        } ?: EntryData.TYPE_ROOT
    }

    private fun makeEntryTag(tag: String, @EntryType type: Int): String {
        return tag + ENTRY_TAG_SEPARATOR + type
    }

    private val isStateSaved: Boolean
        get() = activity.getSupportFragmentManager().isStateSaved

    private fun onFragmentBroughtToFront(fragmentTag: String, parentFragmentTag: String) {
        topFragmentState.value = TopFragmentState(fragmentTag, parentFragmentTag)
    }

    private fun registerBackStackListener() {
        if (backStackListener == null) {
            backStackListener = BackStackListener(this)
            backStackListener?.let {
                activity.getSupportFragmentManager().addOnBackStackChangedListener(it)
            }
        }
    }

    private fun unregisterBackStackListener() {
        backStackListener?.let {
            activity.getSupportFragmentManager().removeOnBackStackChangedListener(it)
            it.onStop()
        }
        backStackListener = null
    }

    private class BackStackListener internal constructor(helper: TransactionHelper) : FragmentManager.OnBackStackChangedListener {
        private var stateRunnable: Runnable? = null
        private var helper: TransactionHelper?

        init {
            this.helper = helper
            // set up initial states
            notifyTopFragment(helper.activity.getSupportFragmentManager())
        }

        override fun onBackStackChanged() {
            helper?.let {
                val manager = it.activity.getSupportFragmentManager()
                val fragment = manager.findFragmentById(R.id.browser)
                notifyTopFragment(manager)
                if (fragment is BrowserScreen) {
                    setBrowserState(shouldKeepBrowserRunning(it), it)
                }
            }
        }

        fun onStop() {
            helper = null
            stateRunnable = null
        }

        private fun shouldKeepBrowserRunning(helper: TransactionHelper): Boolean {
            val manager = helper.activity.getSupportFragmentManager()
            val entryCount = manager.backStackEntryCount
            for (i in entryCount - 1 downTo 0) {
                val entry = manager.getBackStackEntryAt(i)
                if (helper.getEntryType(entry) != EntryData.TYPE_FLOATING) {
                    return false
                }
            }
            return true
        }

        private fun setBrowserState(isForeground: Boolean, helper: TransactionHelper) {
            stateRunnable = Runnable { setBrowserForegroundState(isForeground) }
            val actor = getTopAnimationAccessibleFragment(helper)
            var anim: Animation? = null
            if (actor == null || actor.customEnterTransition.also { anim = it } == null || anim?.hasEnded() == true) {
                executeStateRunnable()
            } else {
                anim?.setAnimationListener(object : Animation.AnimationListener {
                    override fun onAnimationStart(animation: Animation) {}
                    override fun onAnimationEnd(animation: Animation) {
                        executeStateRunnable()
                    }

                    override fun onAnimationRepeat(animation: Animation) {}
                })
            }
        }

        private fun setBrowserForegroundState(isForeground: Boolean) {
            helper?.let {
                val manager = it.activity.getSupportFragmentManager()
                val browserFragment = manager.findFragmentById(R.id.browser) as BrowserScreen
                if (isForeground) {
                    browserFragment.goForeground()
                } else {
                    browserFragment.goBackground()
                }
            }
        }

        private fun executeStateRunnable() {
            stateRunnable?.run()
            stateRunnable = null
        }

        private fun notifyTopFragment(manager: FragmentManager) {
            val entryCount = manager.backStackEntryCount
            var fragmentTag = ""
            if (entryCount > 0) {
                val entry = manager.getBackStackEntryAt(entryCount - 1)
                helper?.let {
                    fragmentTag = it.getEntryTag(entry)
                }
            }
            var parentFragmentTag = ""
            if (entryCount > 1) {
                val entry = manager.getBackStackEntryAt(entryCount - 2)
                helper?.let {
                    parentFragmentTag = it.getEntryTag(entry)
                }
            }
            helper?.onFragmentBroughtToFront(fragmentTag, parentFragmentTag)
        }

        fun getTopAnimationAccessibleFragment(helper: TransactionHelper): FragmentAnimationAccessor? {
            val top = helper.latestCommitFragment
            return if (top != null && top is FragmentAnimationAccessor) {
                top
            } else null
        }
    }

    internal object EntryData {
        /**
         * argument passed to [FragmentTransaction.addToBackStack], pressing back when this
         * type of fragment is in foreground will close the app
         */
        const val TYPE_ROOT = 0

        /**
         * argument passed to [FragmentTransaction.addToBackStack], adding fragment of
         * this type will make browser fragment go to background
         */
        const val TYPE_ATTACHED = 1

        /**
         * argument passed to [FragmentTransaction.addToBackStack], browsing fragment
         * will still be in foreground after adding this type of fragment.
         */
        const val TYPE_FLOATING = 2

        @IntDef(TYPE_ROOT, TYPE_ATTACHED, TYPE_FLOATING)
        internal annotation class EntryType
    }

    data class TopFragmentState(val topFragmentTag: String, val parentFragmentTag: String)

    companion object {
        private const val ENTRY_TAG_SEPARATOR = "#"
    }
}
